src/Serialization.SystemTextJson/GraphQLRequestDeserializer.cs

```diff
@@ -0,0 +1,90 @@
+using GraphQL.Server.Common;
+using System;
+using System.Buffers;
+using System.IO;
+using System.Text.Json;
+using System.Threading.Tasks;
+
+namespace GraphQL.Server.Serialization.SystemTextJson
+{
+    public class GraphQLRequestDeserializer : IGraphQLRequestDeserializer
+    {
+        private readonly JsonSerializerOptions _serializerOptions = new JsonSerializerOptions();
+
+        public GraphQLRequestDeserializer(Action<JsonSerializerOptions> configure)
+        {
+            configure(_serializerOptions);
+        }
+
+        public async Task<GraphQLRequestDeserializationResult> FromBodyAsync(Stream stream, long? contentLength)
+        {
+            int length;
+            byte[] jsonBytes;
+            var sharedArrayPool = ArrayPool<byte>.Shared;
+            bool wasRented = false;
+            if (contentLength.HasValue)
+            {
+                length = (int)contentLength;
+                jsonBytes = sharedArrayPool.Rent(length);
+                wasRented = true;
+                await stream.ReadAsync(jsonBytes, 0, length);
+            }
+            else
+            {
+                using var ms = new MemoryStream();
+                await stream.CopyToAsync(ms).ConfigureAwait(false);
+                jsonBytes = ms.ToArray();
```

Comment:
You can avoid double allocation here by using TryGetBuffer. It'll return the buffer you just wrote.

Reply:
ms.TryGetBuffer()?

Reply:
https://docs.microsoft.com/en-us/dotnet/api/system.io.memorystream.trygetbuffer?view=netframework-4.8

Reply:
Ok

Reply:
Or just GetBuffer?

Reply:
Just playing with this now. `length` here on one of the tests is `256` when the actual request length is `85`. Am I accessing the array correctly with `Array`? If so, how can I get the real length so I can truncate it before the reader kicks in?
```
using var ms = new MemoryStream();
await stream.CopyToAsync(ms).ConfigureAwait(false);
if (!ms.TryGetBuffer(out var streamBytes))
{
    return new GraphQLRequestDeserializationResult();
}
jsonBytes = streamBytes.Array;
length = jsonBytes.Length;
```

Reply:
`streamBytes.Array.Count` seems to be `85`. Maybe that's what I need

Reply:
That works, interestingly if I use `GetBuffer` it returns `byte[]` whereas `TryGetBuffer` outs an `ArraySegment<byte>`.

Reply:
You want to use the entire segment as the Offset might not be 0.

Reply:
I've swapped over to `GetBuffer` here, but when I go to truncate just the bytes I need, I'm using the MemoryStream's `Length`, which seemed to be what I needed, but it's a long, so I can't do `.AsSpan(0, ms.Length)` without having the same issue as above. :P

Reply:
Unless this is the part you mentioned where we could avoid a copy by using `PipeReader`, David, which might be worth doing first in case it removes this issue?

Reply:
It’s not worth complicating. I doubt that in real scenarios this length will ever be exceeded. There is really nothing bad with throwing an `NotSupportedException` if real length > `int.MaxValue`.

Reply:
Yea, happy to do that. It'll be optimal for all cases but the read from stream which will buffer the whole thing just to reject if it was too long. Bit nasty. Will see if David has any suggestions on the thread below for that case.

```
if (requestLength.HasValue && requestLength < MaxSharedBufferSize)
    use array pool
else if (requestLength.HasValue)
    throw unsupported exc
else
    buffer stream into memorystream
    if (memorystream length > int.MaxValue)
        throw unsupported exc
    etc
```

Reply:
No need to check `requestLength < MaxSharedBufferSize`. Read my other comments.